### PR TITLE
SaveDropDownSelectedItem

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

[REVIT-141868](https://jira.autodesk.com/browse/REVIT-141868)
DynamoRevit 2.x Categories Dropdown Produces Inconsistent Results Between Revit versions

I believe this is a more complete version of what you're trying to achieve. @ZiyunShang Let me know how you want to cooperate this task

TODOs:
- [x] Add Json Unit tests
- [x] Add XML Unit tests
- [x] Testing undo redo

```Json
"Nodes": [
    {
      "ConcreteType": "DSRevitNodesUI.Categories, DSRevitNodesUI",
      "SelectedIndex": 6,
      "SelectedString": "OST_DuctTerminal",
      "NodeType": "ExtensionNode",
      "Id": "e16b9ac4ad604e7eb4d29a7975335d7c",
      "Inputs": [],
      "Outputs": [
        {
          "Id": "ac2e1b04ef994a4fb5dbe4b2855d1097",
          "Name": "Category",
          "Description": "The selected Category.",
          "UsingDefaultValue": false,
          "Level": 2,
          "UseLevels": false,
          "KeepListStructure": false
        }
      ],
      "Replication": "Disabled",
      "Description": "All built-in categories."
    }
  ]
```

I have did my manual testing in three basic cases:
- [x] Using master build, place a categories node into canvas and save
- [x] Using master build, opening a DYN saved in Dynamo 2.1.0 w/ R2020 worked
- [x] Using master build, opening a DYN saved in Dynamo 1.3.3 w/ R2019 worked
image

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner @ZiyunShang

### CODE REVIEW

File: src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/DSDropDownBase.cs [hunk 1]
```diff
@@ -40,18 +40,30 @@ public void CustomizeView(DSDropDownBase model, NodeView nodeView)
...
             };
             combo.SetBinding(Selector.SelectedIndexProperty, indexBinding);
+
+            // bind the selected item to the model property SelectedString
+            var itemBinding = new Binding("SelectedString")
+            {
+                Mode = BindingMode.TwoWay,
+                Source = model
+            };
+            combo.SetBinding(Selector.SelectedItemProperty, itemBinding);
```
Comment (QilongTang) [inline src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/DSDropDownBase.cs]: Maybe I can bind to selected value in this case, need some testing

**Comment by ZiyunShang** [comment]
Hi @QilongTang , it seems like all the subclass of DSDropDownBase would set their SelectedString and save to .dyn files, right?  It looks good to me.

**Comment by QilongTang** [comment]
@ZiyunShang I got several cases working, but seems the binding I did is not correct so that selection string instead of item string are saved when you create a new DYN. I may need your help understadning that. Feel free to grab the branch and do some testing. Opening a 1.3 graph has been validated so far.

File: src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/DSDropDownBase.cs (hunk 1 above)
Comment (QilongTang) [reply src/Libraries/CoreNodeModelsWpf/NodeViewCustomizations/DSDropDownBase.cs]: Not needed any more

**Comment by mjkkirschner** [comment]
@QilongTang in your manual testing did you also switch the version of revit or only the version of dynamo on which you opened and saved the graphs? To replicate this the number of items or order of them must changed, which would be accomplished by adding new types (family types) or switching revit version. Seems a unit test could also replicate this issue with some mocked data.

**Comment by QilongTang** [comment]
@mjkkirschner Yes I tried opening a DYN saved w/ Dynamo 1.3.3 Revit 2019 in Dynamo 2.2.0 R2020, will make unit tests

**Comment by mjkkirschner** [comment]
LGTM

**Comment by mjkkirschner** [comment]
@QilongTang this is to cherry pick right?

**Comment by QilongTang** [comment]
@mjkkirschner This is for master branch

**Comment by mjkkirschner** [comment]
@QilongTang you don't intend to get this to 2.1 as well?

**Comment by QilongTang** [comment]
@mjkkirschner For some reason, the self CI job has been hanging so I can't cherry-pick before that comes out

**Pull request merged**